icu4j/main/core/src/main/java/com/ibm/icu/text/UnicodeSet.java

```diff
@@ -5127,5 +5143,179 @@ public static void setDefaultXSymbolTable(XSymbolTable xSymbolTable) {
+
+    private class CodePointsIterator implements Iterator<Integer> {
+        private int pos = 0;
+        private int lastCp = -1; // don't call `charAt` twice, in `hasNext()` and `next()`
+
+        @Override
+        public boolean hasNext() {
+            lastCp = charAt(pos++);
+            return lastCp != -1;
+        }
+
+        @Override
+        public Integer next() {
+            if (lastCp != -1) {
+                return lastCp;
+            } else {
+                throw new NoSuchElementException();
+            }
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class CodePointsSpliterator implements Spliterator.OfInt {
+        /*
+         * WARNING: although it is tempting to also use `Spliterator.SIZED`, it is not safe,
+         * as `UnicodeSet.size()` does not return the number of code points,
+         * it returns code points + strings
+         *
+         * Quote from the `UnicodeSet.size()` doc:
+         *     > Returns the number of elements in this set (its cardinality)
+         *     > Note than the elements of a set may include both individual
+         *     > codepoints and strings."
+         *
+         * And from the Spliterator.SIZED doc:
+         *     > Characteristic value signifying that the value returned from
+         *     > {@code estimateSize()} prior to traversal or splitting represents a
+         *     > finite size that, in the absence of structural source modification,
+         *     > represents an exact count of the number of elements that would be
+         *     > encountered by a complete traversal.
+         * Note "exact count of the number of elements that would be encountered"
+         */
+        private final static int CHARACTERISTICS = Spliterator.ORDERED | Spliterator.DISTINCT | Spliterator.NONNULL;
+
+        private final UnicodeSet uset;
+        private int index;
+
+        CodePointsSpliterator(UnicodeSet uset) {
+            this.uset = uset;
+            this.index = 0;
+        }
+
+        @Override
+        public long estimateSize() {
+            return uset.size();
+        }
+
+        @Override
+        public int characteristics() {
+            // TODO: consider calling `isFrozen()` and also return `Spliterator.IMMUTABLE` if true?
+            return CHARACTERISTICS;
+        }
+
+        @Override
+        public Spliterator.OfInt trySplit() {
+            /* From the doc:
+             *   > This method may return null for any reason, including emptiness, inability to split after
+             *   > traversal has commenced, data structure constraints, and efficiency considerations.
+             */
+            return null;
+        }
+
+        @Override
+        public boolean tryAdvance(IntConsumer action) {
+            if (action == null)
+                throw new NullPointerException();
+            int cp = uset.charAt(index++);
```

Comment:
not charAt()...

Reply:
Done
Refactored to work on the list, not calling `charAt`